packages/edit-navigation/src/store/actions.js

```diff
@@ -42,119 +30,67 @@ export function setSelectedMenuId( menuId ) {
-					url: 'Placeholder',
-					menu_order: 1,
-				},
-			} );
-
-			mapping[ menuItem.id ] = block.clientId;
-			const menuItems = yield resolveMenuItems( menuId );
-			yield dispatch(
-				'core',
-				'receiveEntityRecords',
-				'root',
-				'menuItem',
-				[ ...menuItems, menuItem ],
-				menuItemsQuery( menuId ),
-				false
-			);
-		}
-		stack.push( ...block.innerBlocks );
-	}
-
-	yield {
-		type: 'SET_MENU_ITEM_TO_CLIENT_ID_MAPPING',
-		postId: post.id,
-		mapping,
-	};
-} );
-
 /**
  * Converts all the blocks into menu items and submits a batch request to save everything at once.
  *
  * @param {Object} post A navigation post to process
  * @return {Function} An action creator
  */
-export const saveNavigationPost = serializeProcessing( function* ( post ) {
-	const menuId = post.meta.menuId;
-	const menuItemsByClientId = mapMenuItemsByClientId(
-		yield resolveMenuItems( menuId ),
-		yield getMenuItemToClientIdMapping( post.id )
-	);
-
+export const saveNavigationPost = ( post ) => async ( {
+	registry,
+	dispatch,
+} ) => {
+	const lock = await registry
+		.dispatch( coreDataStore )
+		.__unstableAcquireStoreLock( STORE_NAME, [ 'savingMenu' ], {
+			exclusive: true,
+		} );
 	try {
-		// Save edits to the menu, like the menu name.
-		yield dispatch(
-			'core',
-			'saveEditedEntityRecord',
-			'root',
-			'menu',
-			menuId
-		);
+		const menuId = post.meta.menuId;
 
-		const error = yield select(
-			'core',
-			'getLastEntitySaveError',
-			'root',
-			'menu',
-			menuId
-		);
+		await registry
+			.dispatch( coreDataStore )
+			.saveEditedEntityRecord( 'root', 'menu', menuId );
+
+		const error = registry
+			.select( coreDataStore )
+			.getLastEntitySaveError( 'root', 'menu', menuId );
 
 		if ( error ) {
 			throw new Error( error.message );
 		}
 
-		// Save blocks as menu items.
-		const batchSaveResponse = yield* batchSave(
-			menuId,
-			menuItemsByClientId,
+		// Make sure all the existing menu items are available before proceeding
+		const oldMenuItems = await registry
+			.resolveSelect( coreDataStore )
+			.getMenuItems( { menus: post.meta.menuId, per_page: -1 } );
+
+		const annotatedBlocks = blocksTreeToFlatList(
 			post.blocks[ 0 ]
-		);
+		).filter( ( { block: { name } } ) => isSupportedBlock( name ) );
 
-		if ( ! batchSaveResponse.success ) {
-			throw new Error( batchSaveResponse.data.message );
-		}
+		await dispatch( batchInsertPlaceholderMenuItems( annotatedBlocks ) );
+		await dispatch( batchUpdateMenuItems( annotatedBlocks, menuId ) );
 
-		// Clear "stub" navigation post edits to avoid a false "dirty" state.
-		yield dispatch(
-			'core',
-			'receiveEntityRecords',
-			NAVIGATION_POST_KIND,
-			NAVIGATION_POST_POST_TYPE,
-			[ post ],
-			undefined
+		// Delete menu items
+		const deletedIds = difference(
+			oldMenuItems.map( ( { id } ) => id ),
+			annotatedBlocks.map( ( { block } ) =>
+				getRecordIdFromBlock( block )
+			)
 		);
+		await dispatch( batchDeleteMenuItems( deletedIds ) );
```

Comment:
Why can't deletion happen in the same batch as updating?

Reply:
See https://github.com/WordPress/gutenberg/pull/34541#discussion_r709534744

It is just a little cleaner to delete items after we know that create and update requests have completely successfully.